### DISCUSSION TITLE
How to include pprof when using as framework?

### DISCUSSION THREAD

**Discussion by Plorenzo** [pocketbase/pocketbase#1921] [Q&A]
I would like to add the live [pprof](https://pkg.go.dev/net/http/pprof) endpoint to an app that I'm building using Pocketbase as a framework. 

I've added the import `import _ "net/http/pprof"` but the routes /debug/* are returning 404. 

I'm guess is because she svelte app is capturing all the routes, so I need to[ register a new route ](https://pocketbase.io/docs/router) but not sure how to do a handler that exposes the /debug routes.

**Comment by ganigeorgiev** [comment]
Hm, I think it is conflicting with the trailing slash redirect middleware (or at least the Index handler because of https://cs.opensource.google/go/go/+/refs/tags/go1.20.1:src/net/http/pprof/pprof.go;l=374). One workaround could be to access it with double slash (`http://127.0.0.1:8090/debug/pprof//`), something like:

```go
package main

import (
	"log"
	"net/http"
	"net/http/pprof"

	"github.com/labstack/echo/v5"
	"github.com/pocketbase/pocketbase"
	"github.com/pocketbase/pocketbase/core"
)

func wrapHandlerFunc(h http.HandlerFunc) echo.HandlerFunc {
	return func(c echo.Context) error {
		h.ServeHTTP(c.Response(), c.Request())
		return nil
	}
}

func main() {
	app := pocketbase.New()

	app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
		pprofGroup := e.Router.Group("/debug/pprof")
		
		// access with /debug/pprof//
		pprofGroup.GET("/", wrapHandlerFunc(pprof.Index))

		pprofGroup.GET("/allocs", wrapHandlerFunc(pprof.Handler("allocs").ServeHTTP))
		pprofGroup.GET("/block", wrapHandlerFunc(pprof.Handler("block").ServeHTTP))
		pprofGroup.GET("/cmdline", wrapHandlerFunc(pprof.Cmdline))
		pprofGroup.GET("/goroutine", wrapHandlerFunc(pprof.Handler("goroutine").ServeHTTP))
		pprofGroup.GET("/heap", wrapHandlerFunc(pprof.Handler("heap").ServeHTTP))
		pprofGroup.GET("/mutex", wrapHandlerFunc(pprof.Handler("mutex").ServeHTTP))
		pprofGroup.GET("/profile", wrapHandlerFunc(pprof.Profile))
		pprofGroup.POST("/symbol", wrapHandlerFunc(pprof.Symbol))
		pprofGroup.GET("/symbol", wrapHandlerFunc(pprof.Symbol))
		pprofGroup.GET("/threadcreate", wrapHandlerFunc(pprof.Handler("threadcreate").ServeHTTP))
		pprofGroup.GET("/trace", wrapHandlerFunc(pprof.Trace))

		return nil
	})

	if err := app.Start(); err != nil {
		log.Fatal(err)
	}
}
```

_I'll consider sometime in the future to rework the trailing slash middleware since it is causing other issues with the static files serving (or maybe enabling it only for the `/api/*` router group)._

**Reply by ganigeorgiev** [reply]
Alternatively, as noted in the [pprof readme](https://pkg.go.dev/net/http/pprof) and depending on what you want to profile, you can also start a separate http server:

```go
package main

import (
	"log"
	"net/http"

	"github.com/pocketbase/pocketbase"
	"github.com/pocketbase/pocketbase/core"

	_ "net/http/pprof"
)

func main() {
	app := pocketbase.New()

	app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
		go func() {
			log.Println(http.ListenAndServe("127.0.0.1:6060", nil))
		}()

		return nil
	})

	if err := app.Start(); err != nil {
		log.Fatal(err)
	}
}
```
_(the pprof will be accessable at `http://localhost:6060/debug/pprof/`)_

**Comment by Plorenzo** [comment]
thanks for the suggestions! I'll try and report back
